README
NAME
    WWW::RoboCop - Police your URLs!

VERSION
    version 0.000003

SYNOPSIS
        use feature qw( state );

        use WWW::RoboCop;

        my $robocop = WWW::RoboCop->new(
            is_url_whitelisted => sub {
                state $count = 0;
                return $count++ < 5; # just crawl 5 URLs
            },
        );

        $robocop->crawl( 'http://host.myhost.com/start' );

        my %history = $robocop->get_report;

        # %history = (
        #    'http://myhost.com/one' => { status => 200, ... },
        #    'http://myhost.com/two' => { status => 404, ... },
        #    ...
        # )

DESCRIPTION
    BETA BETA BETA!

    "WWW::RoboCop" is a dead simple, somewhat opinionated robot. Given a
    starting page, this module will crawl only URLs which have been
    whitelisted by the "is_url_whitelisted" callback. It then creates a
    report of all visited pages, keyed on URL. You are encouraged to provide
    your own report creation callback so that you can collect all of the
    information which you require for each URL.

CONSTRUCTOR AND STARTUP
  new()
    Creates and returns a new "WWW::RoboCop" object.

    Below are the arguments which you may pass to "new" when creating an
    object.

   is_url_whitelisted
    This argument is required. You must provide an anonymous subroutine
    which will return true or false based on some arbitrary criteria which
    you provide. The two arguments to this anonymous subroutine will be a
    WWW::Mechanize::Link object as well as the referring URL, in the form of
    a URI object.

    Your sub might look something like this:

        use feature qw( state );

        use URI;
        use WWW::RoboCop;

        my $robocop = WWW::RoboCop->new(
            is_url_whitelisted => sub {
                my $link          = shift;
                my $referring_url = shift;

                my $upper_limit = 100;
                my $host = 'some.host.com';

                state $limit = 0;

                return 0 if $limit > $upper_limit;
                my $uri = URI->new( $link->url_abs );

                # if the referring_url matches the host then this is a 1st degree
                # outbound web link

                if ( $uri->host eq $host || $referring_url->host eq $host ) {
                    ++$limit;
                    return 1;
                }
                return 0;
            }
        );

   report_for_url
    This argument is not required, but is highly recommended. The arguments
    to this anonymous subroutine will be an HTTP::Response object as well as
    the referring URL in the form of a URI object. Your sub might look
    something like this:

        my $reporter = sub {
            my $response      = shift;    # HTTP::Response object
            my $referring_url = shift;    # URI object
            return {
                redirects => [
                    map { +{ status => $_->code, uri => $_->base->as_string } }
                        $res->redirects
                ],
                referrer => $referring_url,
                status   => $res->code,
            };
        };

        my $robocop = WWW::RoboCop->new(
            is_url_whitelisted => sub { ... },
            report_for_url     => $reporter,
        );

    That would give you a HashRef with the status code for each link visited
    (200, 404, 500, etc) as well as the referring URL (the page on which the
    link was found) and a list of any redirects which were followed in order
    to get to this URL.

    The default "report_for_url" sub will already provide something like the
    above, but you should only treat this as a stub method while you get up
    and running. Since it's only meant to be an example, the format of the
    default report could change at some future date without notice. You
    should not rely on or expect it to remain consistent in future. If you
    are going to rely on this module, you should provide your own reporting
    logic.

   ua( WWW::Mechanize )
    You can provide your own UserAgent object to this class. It should be of
    the WWW::Mechanize family. If you're looking for a significant speed
    boost while under development, consider providing a
    WWW::Mechanize::Cached object. This can give you enough of a speedup to
    save you from getting distracted and going off to read Hacker News while
    you wait.

        use CHI;
        use WWW::Mechanize::Cached;
        use WWW::RoboCop;

        my $cache = CHI->new(
            driver => 'File',
            root_dir => /tmp/cache-example',
        );

        my $robocop = WWW::RoboCop->new(
            is_url_whitelisted => sub { ... },
            ua => WWW::Mechanize::Cached->new( cache => $cache ),
        );

    If you're not using a Cached agent, be sure to disable autocheck.

        my $robocop = WWW::RoboCop->new(
            is_url_whitelisted => sub { ... },
            ua => WWW::Mechanize->new( autocheck => 0 ),
        );

  crawl( $url )
    This method sets the "WWW::RoboCop" in motion. The robot will only come
    to a halt once has exhausted all of the whitelisted URLs it can find.

  get_report
    This method returns a Hash of crawling results, keyed on the URLs
    visited. By default, it returns a very simple Hash, containing only the
    status code of the visited URL. You are encouraged to provide your own
    callback so that you can get a detailed report returned to you. You can
    do this by providing a "report_for_url" callback when instantiating the
    object.

    The default report looks something like this:

        # %history = (
        #    'http://myhost.com/one' => { status => 200, ... },
        #    'http://myhost.com/two' => { status => 404, ... },
        # )

    See examples/crawl-host.pl, which is included with this distribution, to
    get a dump of the default report.

AUTHOR
    Olaf Alders <olaf@wundercounter.com>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2015 by MaxMind, Inc..

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)



Changes
Revision history for WWW-RoboCop

{{$NEXT}}

0.000103  2023-11-28 14:26:04Z
    - Check the image links too. (GH#6) (brian d foy)

0.000102  2022-12-02 18:25:14Z
    - Use File::XDG for cache dir in example script (GH#4) (Olaf Alders)

0.000101  2022-07-14 16:16:48Z
    - Make examples/crawl-host.pl more flexible (GH#3) (Olaf Alders)

0.000100  2020-07-08 22:46:45Z
    - Replace 'whitelist' term with 'allowed' (GH#1) (Andy Jack)
      is_url_whitelisted has now been superceded by is_url_allowed. No major
      version bump as this is technically in still in beta.

0.000004  2015-04-17 00:18:09+02:00 Europe/Berlin
    - Adds test dependency

0.000003  2015-04-16 23:46:27+02:00 Europe/Berlin
    - Return URI objects rather than strings in default report
    - Default report only has a redirects key if redirects were followed
    - Add target_url to default report when redirects have been followed

0.000002  2015-01-08 23:31:28-05:00 America/Toronto
    - Don't try to follow href="#"
    - get_report() now returns a Hash rather than HashRef

0.000001  2015-01-07 17:44:31-05:00 America/Toronto
    - Initial release.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.031.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Police your URLs!",
  "AUTHOR" => "Olaf Alders <olaf\@wundercounter.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "WWW-RoboCop",
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "WWW::RoboCop",
  "PREREQ_PM" => {
    "Moo" => 0,
    "MooX::HandlesVia" => 0,
    "MooX::StrictConstructor" => 0,
    "Type::Params" => 0,
    "Types::Standard" => 0,
    "Types::URI" => 0,
    "URI" => 0,
    "WWW::Mechanize" => 0,
    "feature" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Plack::Handler::HTTP::Server::Simple" => "0.016",
    "Plack::Test::Agent" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.000104",
  "test" => {
    "TESTS" => "t/*.t t/WWW/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moo" => 0,
  "MooX::HandlesVia" => 0,
  "MooX::StrictConstructor" => 0,
  "Plack::Handler::HTTP::Server::Simple" => "0.016",
  "Plack::Test::Agent" => 0,
  "Test::More" => 0,
  "Type::Params" => 0,
  "Types::Standard" => 0,
  "Types::URI" => 0,
  "URI" => 0,
  "WWW::Mechanize" => 0,
  "feature" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


examples/crawl-host.pl
#!/usr/bin/env perl

=head1 SYNOPSIS

    perl examples/crawl-host.pl www.somehost.com

=head1 DESCRIPTION

This simple script shows you a dump of the default report.  You're encouraged
to provide your own reporting callback in order to customize your report.

=cut

use strict;
use warnings;
use feature qw( say state );

use CHI;
use Data::Printer filters => { -external => ['URI'], };
use Path::Tiny qw( path );
use WWW::RoboCop;
use WWW::Mechanize::Cached;

my $cache = CHI->new(
    driver   => 'File',
    root_dir => path( '~/.www-robocop-cache' )->stringify,
);

my $host        = shift @ARGV;
my $upper_limit = 10;

die 'usage: perl examples/crawl-host.pl www.somehost.com' unless $host;

my $robocop = WWW::RoboCop->new(
    is_url_allowed => sub {
        my $link          = shift;
        my $referring_url = shift;

        state $limit = 0;

        return 0 if $limit > $upper_limit;
        my $uri = URI->new( $link->url_abs );

       # If the link URI does not match the host but the referring_url matches
       # the host, then this is a 1st degree outbound link.  We'll fetch the
       # page in order to log the status code etc, but we won't index any of
       # the links on it.

        if ( $uri->host eq $host || $referring_url->host eq $host ) {
            ++$limit;
            return 1;
        }
        return 0;
    },
    ua => WWW::Mechanize::Cached->new( cache => $cache ),
);

$robocop->crawl( "http://$host" );

my %report = $robocop->get_report;

p( %report );


lib/WWW/RoboCop.pm
use strict;
use warnings;
use feature qw( state );

package WWW::RoboCop;
our $VERSION = '0.000104';

use Moo;

use MooX::HandlesVia;
use MooX::StrictConstructor;
use Type::Params    qw( compile );
use Types::Standard qw( CodeRef HashRef InstanceOf );
use Types::URI      qw( Uri );
use URI             ();
use WWW::Mechanize  ();

has is_url_allowed => (
    is          => 'ro',
    isa         => CodeRef,
    handles_via => 'Code',
    handles     => { _should_follow_link => 'execute' },
    required    => 1,
);

has report_for_url => (
    is          => 'ro',
    isa         => CodeRef,
    handles_via => 'Code',
    handles     => { _log_response => 'execute' },
    default     => sub {
        sub {
            my $response      = shift;    # HTTP::Response object
            my $referring_url = shift;    # URI object
            return {
                $response->redirects
                ? (
                    redirects => [
                        map {
                            { status => $_->code, uri => $_->request->uri, }
                        } $response->redirects
                    ],
                    target_url => $response->request->uri,
                    )
                : (),
                referrer => $referring_url,
                status   => $response->code,
            };
        };
    },
);

has ua => (
    is      => 'ro',
    isa     => InstanceOf ['WWW::Mechanize'],
    default => sub {
        WWW::Mechanize->new( autocheck => 0 );
    },
);

has _history => (
    is          => 'ro',
    isa         => HashRef,
    handles_via => 'Hash',
    handles     => {
        _add_url_to_history => 'set',
        _has_processed_url  => 'exists',
    },
    init_arg => undef,
    default  => sub { +{} },
);

sub _get {
    my $self          = shift;
    my $fetched_url   = shift;
    my $referring_url = shift;

    my $response = $self->ua->get($fetched_url);
    my $report   = $self->_log_response( $response, $referring_url );
    $self->_add_url_to_history( $fetched_url, $report );

    my @links = map {
        my $u = URI->new( $_->url_abs );
        $u->fragment(undef);
        [ $_, $u ]
        }
        grep { $_->url !~ /\A#/ }                           # no named anchors
        map { @{ $self->ua->$_() } } qw( links images );

    foreach my $tuple (@links) {
        my ( $mech_link, $uri ) = @$tuple;
        next unless $self->_should_follow_link( $mech_link, $fetched_url );
        next if $self->_has_processed_url($uri);
        next unless $uri->can('host');    # no mailto: links
        $self->_get( $uri, $fetched_url );
    }
}

sub crawl {
    my $self = shift;

    state $check = compile(Uri);
    my ($url) = $check->(@_);

    $self->_get($url);
}

sub get_report {
    my $self = shift;
    return %{ $self->_history };
}

1;

__END__

# ABSTRACT: Police your URLs!

=pod

=head1 DESCRIPTION

C<WWW::RoboCop> is a simple, somewhat opinionated robot.  Given a starting
page, this module will crawl only URLs which have been allowed by the
C<is_url_allowed> callback.  It then creates a report of all visited pages,
keyed on URL.  You are encouraged to provide your own report creation callback
so that you can collect all of the information which you require for each URL.

=head1 SYNOPSIS

    use feature qw( state );

    use WWW::RoboCop ();

    my $robocop = WWW::RoboCop->new(
        is_url_allowed => sub {
            state $count = 0;
            return $count++ < 5; # just crawl 5 URLs
        },
    );

    $robocop->crawl( 'https://example.com' );

    my %history = $robocop->get_report;

    # %history = (
    #    'http://myhost.com/one' => { status => 200, ... },
    #    'http://myhost.com/two' => { status => 404, ... },
    #    ...
    # )

See C<examples/crawl-host.pl>, which is included with this distribution, to get a
quick start.

=head1 CONSTRUCTOR AND STARTUP

=head2 new()

Creates and returns a new C<WWW::RoboCop> object.

Below are the arguments which you may pass to C<new> when creating an object.

=head3 is_url_allowed

This argument is required.  You must provide an anonymous subroutine which will
return true or false based on some arbitrary criteria which you provide.  The
two arguments to this anonymous subroutine will be a L<WWW::Mechanize::Link>
object as well as the referring URL, in the form of a L<URI> object.

Your sub might look something like this:

    use feature qw( state );

    use URI ();
    use WWW::RoboCop ();

    my $robocop = WWW::RoboCop->new(
        is_url_allowed => sub {
            my $link          = shift;
            my $referring_url = shift;

            my $upper_limit = 100;
            my $host = 'some.host.com';

            state $limit = 0;

            return 0 if $limit > $upper_limit;
            my $uri = URI->new( $link->url_abs );

            # if the referring_url matches the host then this is a 1st degree
            # outbound web link

            if ( $uri->host eq $host || $referring_url->host eq $host ) {
                ++$limit;
                return 1;
            }
            return 0;
        }
    );

=head3 report_for_url

This argument is not required, but is highly recommended. The arguments to this
anonymous subroutine will be an L<HTTP::Response> object as well as the
referring URL in the form of a L<URI> object.  Your sub might look something
like this:

    my $reporter = sub {
        my $response      = shift;    # HTTP::Response object
        my $referring_url = shift;    # URI object
        return {
            redirects => [
                map { +{ status => $_->code, uri => $_->base->as_string } }
                    $res->redirects
            ],
            referrer => $referring_url,
            status   => $res->code,
        };
    };

    my $robocop = WWW::RoboCop->new(
        is_url_allowed => sub { ... },
        report_for_url     => $reporter,
    );

That would give you a HashRef with the status code for each link visited (200,
404, 500, etc) as well as the referring URL (the page on which the link was
found) and a list of any redirects which were followed in order to get to this
URL.

The default C<report_for_url> sub will already provide something like the
above, but you should only treat this as a stub method while you get up and
running.  Since it's only meant to be an example, the format of the default
report could change at some future date without notice.  You should not rely on
or expect it to remain consistent in future.  If you are going to rely on this
module, you should provide your own reporting logic.

=head3 ua( WWW::Mechanize )

You can provide your own UserAgent object to this class.  It should be of the
L<WWW::Mechanize> family.  If you're looking for a significant speed boost
while under development, consider providing a L<WWW::Mechanize::Cached> object.
This can give you enough of a speedup to save you from getting distracted
and going off to read Hacker News while you wait.

    use CHI ();
    use WWW::Mechanize::Cached ();
    use WWW::RoboCop ();

    my $cache = CHI->new(
        driver => 'File',
        root_dir => /tmp/cache-example',
    );

    my $robocop = WWW::RoboCop->new(
        is_url_allowed => sub { ... },
        ua => WWW::Mechanize::Cached->new( cache => $cache ),
    );

If you're not using a Cached agent, be sure to disable autocheck.

    my $robocop = WWW::RoboCop->new(
        is_url_allowed => sub { ... },
        ua => WWW::Mechanize->new( autocheck => 0 ),
    );

=head2 crawl( $url )

This method sets the C<WWW::RoboCop> in motion.  The robot will only come to a
halt once has exhausted all of the allowed URLs it can find.

=head2 get_report

This method returns a Hash of crawling results, keyed on the URLs visited.
By default, it returns a very simple Hash, containing only the status code
of the visited URL.  You are encouraged to provide your own callback so that
you can get a detailed report returned to you.  You can do this by providing a
C<report_for_url> callback when instantiating the object.

The default report looks something like this:

    # %history = (
    #    'http://myhost.com/one' => { status => 200, ... },
    #    'http://myhost.com/two' => { status => 404, ... },
    # )

See C<examples/crawl-host.pl>, which is included with this distribution, to get
a dump of the default report.

=cut
